session/portforward.go

```diff
@@ -0,0 +1,276 @@
+		if err != nil {
+			return err
+		}
+		f.conn = conn
+	} else {
+		return errors.New("unknown protocol: " + protocol)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	f.ctx = ctx
+	f.ctxCancel = cancelFunc
+
+	go f.Read()
+
+	return nil
+}
+
+func (f *MenderPortForwarder) Close() error {
+	log.Debugf("port-forward[%s/%s] close", f.SessionID, f.ConnectionID)
+	f.ctxCancel()
+	return f.conn.Close()
+}
+
+func (f *MenderPortForwarder) Read() {
+	errChan := make(chan error)
+	dataChan := make(chan []byte)
+
+	go func() {
+		data := make([]byte, readBuffLength)
+
+		for {
+			n, err := f.conn.Read(data)
+			if err != nil {
+				errChan <- err
+				break
+			}
+			if n > 0 {
+				dataChan <- data[:n]
+			}
+		}
+	}()
+
+	for {
+		select {
+		case err := <-errChan:
+			if err != io.EOF {
+				log.Errorf("port-forward[%s/%s] error: %v\n", f.SessionID, f.ConnectionID, err.Error())
+			}
+			// send the stop message
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypePortForward,
+					MsgType:   wspf.MessageTypePortForwardStop,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						wspf.PropertyPortForwardConnectionID: f.ConnectionID,
+					},
+				},
+			}
+			if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+				log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+			}
+			// close the forwarder
+			f.Close()
+		case data := <-dataChan:
+			log.Debugf("port-forward[%s/%s] read %d bytes", f.SessionID, f.ConnectionID, len(data))
+
+			m := &ws.ProtoMsg{
+				Header: ws.ProtoHdr{
+					Proto:     ws.ProtoTypePortForward,
+					MsgType:   wspf.MessageTypePortForward,
+					SessionID: f.SessionID,
+					Properties: map[string]interface{}{
+						wspf.PropertyPortForwardConnectionID: f.ConnectionID,
+					},
+				},
+				Body: data,
+			}
+			if err := f.ResponseWriter.WriteProtoMsg(m); err != nil {
+				log.Errorf("portForwardHandler: webSock.WriteMessage(%+v)", err)
+			}
+		case <-f.ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *MenderPortForwarder) Write(body []byte) error {
+	log.Debugf("port-forward[%s/%s] write %d bytes", f.SessionID, f.ConnectionID, len(body))
+	_, err := f.conn.Write(body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func PortForward() Constructor {
+	f := HandlerFunc(portForwardHandler)
+	return func() SessionHandler { return f }
+}
+
+func portForwardHandler(msg *ws.ProtoMsg, w ResponseWriter) {
+	var err error
+	switch msg.Header.MsgType {
+	case wspf.MessageTypePortForwardNew:
+		err = portForwardHandlerNew(msg, w)
+	case wspf.MessageTypePortForwardStop:
+		err = portForwardHandlerStop(msg, w)
+	case wspf.MessageTypePortForward:
+		err = portForwardHandlerForward(msg, w)
+	default:
+		err = errPortForwardUnkonwnMessageType
+	}
+	if err != nil {
+		log.Errorf("portForwardHandler(%+v)", err)
+
+		response := &ws.ProtoMsg{
+			Header: ws.ProtoHdr{
+				Proto:     ws.ProtoTypeControl,
```

Comment:
The handler cannot speak on behalf of another protocol.
```suggestion
				Proto:     ws.ProtoTypePortForward,
```